File: app/src/main/java/com/andysklyarov/finnotify/framework/soap/SoapCbrUtils.java
package com.andysklyarov.finnotify.framework.soap;

import androidx.annotation.NonNull;

import com.andysklyarov.finnotify.domain.CurrencyInRub;
import com.andysklyarov.finnotify.framework.database.CurrencyDao;
import com.andysklyarov.finnotify.framework.database.CurrencyOnDate;
import com.andysklyarov.finnotify.framework.database.DbUtils;
import com.andysklyarov.finnotify.framework.database.LastDate;
import com.andysklyarov.finnotify.framework.soap.GetCursOnDateXML.request.RequestGetCursOnDateXMLBody;
import com.andysklyarov.finnotify.framework.soap.GetCursOnDateXML.request.RequestGetCursOnDateXMLData;
import com.andysklyarov.finnotify.framework.soap.GetCursOnDateXML.request.RequestGetCursOnDateXMLEnvelope;
import com.andysklyarov.finnotify.framework.soap.GetCursOnDateXML.response.ResponseGetCursOnDateXMLBody;
import com.andysklyarov.finnotify.framework.soap.GetCursOnDateXML.response.ResponseGetCursOnDateXMLEnvelope;
import com.andysklyarov.finnotify.framework.soap.GetCursOnDateXML.response.ResponseGetCursOnDateXMLResponse;
import com.andysklyarov.finnotify.framework.soap.GetCursOnDateXML.response.ResponseGetCursOnDateXMLResult;
import com.andysklyarov.finnotify.framework.soap.GetCursOnDateXML.response.ResponseValuteCursOnDate;
import com.andysklyarov.finnotify.framework.soap.GetCursOnDateXML.response.ResponseValuteData;
import com.andysklyarov.finnotify.framework.soap.GetLatestDateTime.request.RequestLatestDateTimeBody;
import com.andysklyarov.finnotify.framework.soap.GetLatestDateTime.request.RequestLatestDateTimeData;
import com.andysklyarov.finnotify.framework.soap.GetLatestDateTime.request.RequestLatestDateTimeEnvelope;
import com.andysklyarov.finnotify.framework.soap.GetLatestDateTime.response.ResponseLatestDateTimeBody;
import com.andysklyarov.finnotify.framework.soap.GetLatestDateTime.response.ResponseLatestDateTimeData;
import com.andysklyarov.finnotify.framework.soap.GetLatestDateTime.response.ResponseLatestDateTimeEnvelope;

import org.jetbrains.annotations.Nullable;

import java.time.LocalDate;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

import io.reactivex.Single;

import static java.lang.Float.parseFloat;

public final class SoapCbrUtils {

    private static final String INPUT_DATE_TIME_PATTERN = "yyyy-MM-dd";
    private static final String TIME_PATTERN = "T00:00:00";
    private static final String OUTPUT_DATE_TIME_PATTERN = "yyyy-MM-dd'T'HH:mm:ss";
    private static final String INTERNAL_DATE_PATTERN = "yyyyMMdd";

    public static Single<LocalDate> getLastServerDate() {

        RequestLatestDateTimeEnvelope envelope = new RequestLatestDateTimeEnvelope();
        RequestLatestDateTimeBody body = new RequestLatestDateTimeBody();
        RequestLatestDateTimeData data = new RequestLatestDateTimeData();

        body.setLatestDateTimeData(data);
        envelope.setBody(body);

        return HttpUtils.getApiService()
                .getLatestDateTime(envelope)
                .doOnSuccess(response -> writeLastDateToDb(response))
                .onErrorReturn(throwable -> getLastDateFromDb(throwable))
                .map(response -> {
                    String dateTime = response.getBody()
                            .getLatestDateTimeData()
                            .getDateTime();

                    DateTimeFormatter outputFormatter = DateTimeFormatter.ofPattern(OUTPUT_DATE_TIME_PATTERN);
                    return LocalDate.parse(dateTime, outputFormatter); //"2020-04-30T00:00:00"
                });
    }

    public static Single<CurrencyInRub> getCurrencyOnDate(String currencyName, LocalDate date) {
        RequestGetCursOnDateXMLEnvelope envelope = new RequestGetCursOnDateXMLEnvelope();
        RequestGetCursOnDateXMLBody body = new RequestGetCursOnDateXMLBody();
        RequestGetCursOnDateXMLData data = new RequestGetCursOnDateXMLData();

        String formattedDate = getInputDateTimeString(date);

        data.setDateTime(formattedDate);
        body.setGetCursOnDateXML(data);
        envelope.setBody(body);

        return HttpUtils.getApiService()
                .getCursOnDateXML(envelope)
                .doOnSuccess(response -> writeCurrencyToDb(formattedDate, response))
                .onErrorReturn(throwable -> getCurrencyFromDbOrNull(date, throwable))
                .map(response -> getCurrencyInRub(currencyName, response));
    }

    private static void writeLastDateToDb(ResponseLatestDateTimeEnvelope response) {
        CurrencyDao dao = DbUtils.getDatabase().getCurrencyDao();

        String dateTime = response.getBody()
                .getLatestDateTimeData()
                .getDateTime();

        dao.insertLastDate(new LastDate(0, dateTime));
    }

    private static ResponseLatestDateTimeEnvelope getLastDateFromDb(Throwable throwable) {
        if (hasNetworkException(throwable)) {
            CurrencyDao dao = DbUtils.getDatabase().getCurrencyDao();

            ResponseLatestDateTimeEnvelope response = new ResponseLatestDateTimeEnvelope();
            ResponseLatestDateTimeBody body = new ResponseLatestDateTimeBody();
            ResponseLatestDateTimeData data = new ResponseLatestDateTimeData();

            String date = dao.getLastDate().getDate();

            data.setDateTime(date);
            body.setLatestDateTimeData(data);
            response.setBody(body);

            return response;
        } else {
            return null;
        }
    }

    @NonNull
    private static String getInputDateTimeString(LocalDate date) {
        return getFormattedDateString(date, INPUT_DATE_TIME_PATTERN) + TIME_PATTERN;
    }

    @NonNull
    private static String getFormattedDateString(LocalDate date, String format) {
        DateTimeFormatter inputFormatter = DateTimeFormatter.ofPattern(format);
        return inputFormatter.format(date);
    }

    private static void writeCurrencyToDb(String formattedDate, ResponseGetCursOnDateXMLEnvelope response) {
        CurrencyDao dao = DbUtils.getDatabase().getCurrencyDao();

        List<ResponseValuteCursOnDate> currencies = response.getBody()
                .getCursOnDateXMLResponse()
                .getGetCursOnDateXMLResult()
                .getValuteData()
                .getValuteCurses();

        for (int i = 0, size = currencies.size(); i < size; i++) {
            ResponseValuteCursOnDate currentCurs = currencies.get(i);
            CurrencyOnDate currencyOnDate = new CurrencyOnDate(
                    i,
                    currentCurs.valName,
                    currentCurs.nom,
                    currentCurs.curs,
                    currentCurs.code,
                    currentCurs.chCode,
                    formattedDate
            );
            dao.insertCurrencyOnDate(currencyOnDate);
        }
    }

    @Nullable
    private static ResponseGetCursOnDateXMLEnvelope getCurrencyFromDbOrNull(LocalDate date, Throwable throwable) {
        if (hasNetworkException(throwable)) {
            CurrencyDao dao = DbUtils.getDatabase().getCurrencyDao();
            List<CurrencyOnDate> currencies = dao.getCurrenciesOnDate(getInputDateTimeString(date));

            ResponseGetCursOnDateXMLEnvelope respEnvelope = new ResponseGetCursOnDateXMLEnvelope();
            ResponseGetCursOnDateXMLBody respBody = new ResponseGetCursOnDateXMLBody();
            ResponseGetCursOnDateXMLResponse respResponse = new ResponseGetCursOnDateXMLResponse();
            ResponseGetCursOnDateXMLResult respResult = new ResponseGetCursOnDateXMLResult();
            ResponseValuteData respValuteData = new ResponseValuteData();

            List<ResponseValuteCursOnDate> respValuteCursOnDateList = new ArrayList<>();

            for (CurrencyOnDate currencyDb : currencies) {
                ResponseValuteCursOnDate value = new ResponseValuteCursOnDate(
                        currencyDb.getValName(),
                        currencyDb.getNom(),
                        currencyDb.getCurs(),
                        currencyDb.getCode(),
                        currencyDb.getChCode()
                );
                respValuteCursOnDateList.add(value);
            }

            String internalFormattedDate = getFormattedDateString(date, INTERNAL_DATE_PATTERN);
            respValuteData.setOnDate(internalFormattedDate);

            respValuteData.setValuteCurses(respValuteCursOnDateList);
            respResult.setValuteData(respValuteData);
            respResponse.setGetCursOnDateXMLResult(respResult);
            respBody.setCursOnDateXMLResponse(respResponse);
            respEnvelope.setBody(respBody);

            return respEnvelope;
        } else {
            return null;
        }
    }

    @NonNull
    private static CurrencyInRub getCurrencyInRub(String currencyName, ResponseGetCursOnDateXMLEnvelope response) {
        ResponseValuteData responseValuteData = response.getBody()
                .getCursOnDateXMLResponse()
                .getGetCursOnDateXMLResult()
                .getValuteData();

        List<ResponseValuteCursOnDate> currencies = responseValuteData.getValuteCurses();
        String onDateStr = responseValuteData.getOnDate();

        DateTimeFormatter outputFormatter = DateTimeFormatter.ofPattern(INTERNAL_DATE_PATTERN); //"20200901"
        LocalDate localDate = LocalDate.parse(onDateStr, outputFormatter);

        float currencyValueInRub = 0;

        if (currencies != null) {
            for (ResponseValuteCursOnDate currency : currencies) {
                if (currency.chCode.equals(currencyName)) {
                    currencyValueInRub = parseFloat(currency.curs);
                    break;
                }
            }
        }

        return new CurrencyInRub(currencyName, localDate, currencyValueInRub);
    }

    private static boolean hasNetworkException(Throwable throwable) {
        return HttpUtils.NETWORK_EXCEPTIONS.contains(throwable.getClass());
    }
}

File: app/src/main/java/com/andysklyarov/finnotify/framework/soap/SoapCbrUtils.kt
package com.andysklyarov.finnotify.framework.soap

import com.andysklyarov.finnotify.domain.CurrencyInRub
import com.andysklyarov.finnotify.domain.CurrencyName
import com.andysklyarov.finnotify.framework.database.CurrencyDao
import com.andysklyarov.finnotify.framework.database.CurrencyDatabase
import com.andysklyarov.finnotify.framework.database.CurrencyOnDate
import com.andysklyarov.finnotify.framework.database.LastDate
import com.andysklyarov.finnotify.framework.soap.GetCursOnDateXML.*
import com.andysklyarov.finnotify.framework.soap.GetLatestDateTime.*
import io.reactivex.Single
import org.simpleframework.xml.Serializer
import org.simpleframework.xml.core.Persister
import retrofit2.converter.simplexml.SimpleXmlConverterFactory
import java.io.ByteArrayOutputStream
import java.io.OutputStream
import java.time.LocalDate
import java.time.format.DateTimeFormatter

object SoapCbrUtils {
    private const val INPUT_DATE_TIME_PATTERN = "yyyy-MM-dd"
    private const val TIME_PATTERN = "T00:00:00"
    private const val OUTPUT_DATE_TIME_PATTERN = "yyyy-MM-dd'T'HH:mm:ss"
    private const val INTERNAL_DATE_PATTERN = "yyyyMMdd"

    fun getLastServerDate(): Single<LocalDate> {
        val data = RequestLatestDateTimeData()
        val body = RequestLatestDateTimeBody(data)
        val envelope = RequestLatestDateTimeEnvelope(body)

        return HttpUtils.soapApi.getLatestDateTime(envelope)
            .doOnSuccess { response -> writeLastDateToDb(response) }
            .onErrorReturn { throwable -> getLastDateFromDb(throwable) }
            .map { response ->
                val dateTime = response.body?.latestDateTimeDataResponse?.dateTime
                val outputFormatter = DateTimeFormatter.ofPattern(OUTPUT_DATE_TIME_PATTERN)
                LocalDate.parse(dateTime, outputFormatter)
            }
    }

    fun getCurrencyOnDate(currencyCode: String, date: LocalDate): Single<CurrencyInRub> {
        val formattedDate = getInputDateTimeString(date)
        val data = RequestGetCursOnDateXMLData(formattedDate)
        val body = RequestGetCursOnDateXMLBody(data)
        val envelope = RequestGetCursOnDateXMLEnvelope(body)

        return HttpUtils.soapApi
            .getCursOnDateXML(envelope)
            .doOnSuccess { response -> writeCurrencyToDb(formattedDate, response) }
            .onErrorReturn { throwable -> getCurrencyFromDbOrNull(date, throwable) }
            .map { response -> getCurrencyInRub(currencyCode, response) }
    }

    private fun writeLastDateToDb(response: ResponseLatestDateTimeEnvelope) {
        val dao: CurrencyDao = CurrencyDatabase.getDatabase().getCurrencyDao()
        val dateTime = response.body?.latestDateTimeDataResponse?.dateTime
        dao.insertLastDate(LastDate(0, dateTime!!))
    }

    private fun getLastDateFromDb(throwable: Throwable): ResponseLatestDateTimeEnvelope? {
        return if (hasNetworkException(throwable)) {
            val dao: CurrencyDao = CurrencyDatabase.getDatabase().getCurrencyDao()

            val date = dao.getLastDate().date
            val data = ResponseLatestDateTimeData(date)
            val body = ResponseLatestDateTimeBody(data)
            val response = ResponseLatestDateTimeEnvelope(body)

            response
        } else {
            null
        }
    }

    private fun getInputDateTimeString(date: LocalDate): String {
        return getFormattedDateString(date, INPUT_DATE_TIME_PATTERN) + TIME_PATTERN
    }

    private fun getFormattedDateString(date: LocalDate, format: String): String {
        return DateTimeFormatter.ofPattern(format).format(date)
    }

    private fun writeCurrencyToDb(
        formattedDate: String,
        response: ResponseGetCursOnDateXMLEnvelope
    ) {
        val dao: CurrencyDao = CurrencyDatabase.getDatabase().getCurrencyDao()
        val currencies = response.body
            ?.cursOnDateXMLResponse
            ?.getCursOnDateXMLResult
            ?.valuteData
            ?.valuteCurses

        for (i in currencies!!.indices) {
            val (valName, nom, curs, code, chCode) = currencies[i]
            val currencyOnDate = CurrencyOnDate(
                i.toLong(), valName, nom, curs, code, chCode, formattedDate
            )
            dao.insertCurrencyOnDate(currencyOnDate)
        }
    }

    private fun getCurrencyFromDbOrNull(
        date: LocalDate,
        throwable: Throwable
    ): ResponseGetCursOnDateXMLEnvelope? {
        return if (hasNetworkException(throwable)) {
            val dao: CurrencyDao = CurrencyDatabase.getDatabase().getCurrencyDao()
            val currencies = dao.getCurrenciesOnDate(getInputDateTimeString(date))

            val respValuteCursOnDateList = ArrayList<ResponseValuteCursOnDate>()

            for (currencyDb in currencies) {
                val value = ResponseValuteCursOnDate(
                    currencyDb.valName,
                    currencyDb.nom,
                    currencyDb.curs,
                    currencyDb.code,
                    currencyDb.chCode
                )
                respValuteCursOnDateList.add(value);
            }

            val internalFormattedDate = getFormattedDateString(date, INTERNAL_DATE_PATTERN)

            val respValuteData = ResponseValuteData(respValuteCursOnDateList, internalFormattedDate)
            val respResult = ResponseGetCursOnDateXMLResult(respValuteData)
            val respResponse = ResponseGetCursOnDateXMLResponse(respResult)
            val respBody = ResponseGetCursOnDateXMLBody(respResponse)
            val respEnvelope = ResponseGetCursOnDateXMLEnvelope(respBody)

            respEnvelope
        } else {
            null
        }
    }

    private fun getCurrencyInRub(
        currencyCode: String,
        response: ResponseGetCursOnDateXMLEnvelope
    ): CurrencyInRub {

        val responseValuteData =
            response.body?.cursOnDateXMLResponse?.getCursOnDateXMLResult?.valuteData

        val currencies = responseValuteData?.valuteCurses
        val onDateStr = responseValuteData?.onDate

        val outputFormatter = DateTimeFormatter.ofPattern(INTERNAL_DATE_PATTERN); //"20200901"
        val localDate = LocalDate.parse(onDateStr, outputFormatter);


        var currencyValueInRub = 0.0f;
        var denomination = 0;
        var currencyName = CurrencyName("Empty", "Empty");

        for (currency in currencies!!) {
            if (currency.chCode == currencyCode) {
                currencyValueInRub = currency.curs.toFloat()
                denomination = currency.nom.toInt()
                currencyName = CurrencyName(currency.valName.trim(), currency.chCode)
            }
        }

        return CurrencyInRub(currencyName, localDate, denomination, currencyValueInRub)
    }

    private fun hasNetworkException(throwable: Throwable): Boolean {
        return HttpUtils.NETWORK_EXCEPTIONS.contains(throwable.javaClass)
    }

}